File: app/src/test/java/pl/srw/billcalculator/history/HistoryPresenterTest.java
package pl.srw.billcalculator.history;

import android.view.View;

import org.greenrobot.greendao.query.LazyList;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Arrays;

import junitparams.JUnitParamsRunner;
import junitparams.Parameters;
import pl.srw.billcalculator.Whitebox;
import pl.srw.billcalculator.bill.SavedBillsRegistry;
import pl.srw.billcalculator.db.Bill;
import pl.srw.billcalculator.db.History;
import pl.srw.billcalculator.db.PgeG11Bill;
import pl.srw.billcalculator.history.list.item.HistoryViewItem;
import pl.srw.billcalculator.persistence.type.BillType;
import pl.srw.billcalculator.settings.global.SettingsRepo;
import pl.srw.billcalculator.type.Provider;
import pl.srw.billcalculator.util.BillSelection;
import pl.srw.billcalculator.wrapper.HistoryRepo;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(JUnitParamsRunner.class)
public class HistoryPresenterTest {

    @InjectMocks private HistoryPresenter sut;
    
    @Mock HistoryPresenter.HistoryView view;
    @Mock SettingsRepo settings;
    @Mock HistoryRepo history;
    @Mock SavedBillsRegistry savedBillsRegistry;
    @Mock LazyList<History> listData;
    @Mock BillSelection selection;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        Whitebox.setInternalState(sut, "view", view);
        when(history.getAll()).thenReturn(listData);
    }

    @Test
    public void onFirstBind_whenFirstLaunch_showsWelcomeDialog() throws Exception {
        // GIVEN
        when(settings.isFirstLaunch()).thenReturn(true);

        // WHEN
        sut.onFirstBind();

        // THEN
        verify(view).showWelcomeDialog();
    }

    @Test
    public void onFirstBind_whenNotFirstLaunch_dontShowsWelcomeDialog() throws Exception {
        // GIVEN
        when(settings.isFirstLaunch()).thenReturn(false);

        // WHEN
        sut.onFirstBind();

        // THEN
        verify(view, never()).showWelcomeDialog();
    }

    @Test
    public void onFirstBind_fetchAllHistory() throws Exception {
        // WHEN
        sut.onFirstBind();

        // THEN
        verify(history).getAll();
    }

    @Test
    public void onFirstBind_setsHistoryDataOnList() throws Exception {
        // WHEN
        sut.onFirstBind();

        // THEN
        verify(view).setListData(listData);
        verify(view).redrawList();
    }

    @Test
    public void onNewViewRestoreState_whenNeedRefresh_fetchAllHistory() throws Exception {
        // GIVEN
        Whitebox.setInternalState(sut, "needRefresh", true);

        // WHEN
        sut.onNewViewRestoreState();

        // THEN
        verify(history).getAll();
    }

    @Test
    public void onNewViewRestoreState_setsHistoryDataOnList() throws Exception {
        // GIVEN
        final LazyList<History> list = mock(LazyList.class);
        Whitebox.setInternalState(sut, "historyData", list);

        // WHEN
        sut.onNewViewRestoreState();

        // THEN
        verify(view).setListData(list);
        verify(view).redrawList();
    }

    @Test
    public void helpMenuClicked_showsHelp() throws Exception {
        // WHEN
        sut.helpMenuClicked();

        // THEN
        verify(view).showHelp();
    }

    @Test
    public void settingsClicked_opensSettings() throws Exception {
        // WHEN
        sut.settingsClicked();

        // THEN
        verify(view).openSettings();
    }

    @Test
    public void settingsClicked_closesDrawer() throws Exception {
        // WHEN
        sut.settingsClicked();

        // THEN
        verify(view).closeDrawer();
    }

    @Test
    public void handleBackPressed_whenDrawerOpened_closesDrawer() throws Exception {
        // GIVEN
        when(view.isDrawerOpen()).thenReturn(true);

        // WHEN
        sut.handleBackPressed();

        // THEN
        verify(view).closeDrawer();
    }

    @Test
    public void handleBackPressed_whenDrawerOpened_returnsTrue() throws Exception {
        // GIVEN
        when(view.isDrawerOpen()).thenReturn(true);

        // WHEN
        boolean result = sut.handleBackPressed();

        // THEN
        assertTrue(result);
    }

    @Test
    public void handleBackPressed_whenDrawerClosed_returnsFalse() throws Exception {
        // GIVEN
        when(view.isDrawerOpen()).thenReturn(false);

        // WHEN
        boolean result = sut.handleBackPressed();

        // THEN
        assertFalse(result);
    }

    @Test
    public void historyClicked_closesDrawer() throws Exception {
        // WHEN
        sut.historyClicked();

        // THEN
        verify(view).closeDrawer();
    }

    @Test
    @Parameters({"PGE", "PGNIG", "TAURON"})
    public void newBillClicked_showsForm(Provider provider) throws Exception {
        // WHEN
        sut.newBillClicked(provider);

        // THEN
        verify(view).showNewBillForm(provider);
    }

    @Test
    public void aboutClicked_showsAbout() throws Exception {
        // WHEN
        sut.aboutClicked();

        // THEN
        verify(view).showAbout();
    }

    @Test
    public void aboutClicked_closesDrawer() throws Exception {
        // WHEN
        sut.aboutClicked();

        // THEN
        verify(view).closeDrawer();
    }

    @Test
    public void onListItemDismissed_deleteBillWithPrices() throws Exception {
        // GIVEN
        Bill bill = mock(PgeG11Bill.class);
        final Long id = 1L;
        when(bill.getId()).thenReturn(id);
        final Long pricesId = 2L;
        when(bill.getPricesId()).thenReturn(pricesId);

        // WHEN
        sut.onListItemDismissed(0, bill);

        // THEN
        verify(history).deleteBillWithPrices(BillType.PGE_G11, id, pricesId);
    }

    @Test
    public void onListItemDismissed_fetchAllHistory() throws Exception {
        // GIVEN
        final Bill bill = mock(PgeG11Bill.class);

        // WHEN
        sut.onListItemDismissed(0, bill);

        // THEN
        verify(history).getAll();
    }

    @Test
    public void onListItemDismissed_viewRemovesItemFromList() throws Exception {
        // GIVEN
        final int position = 1;
        final Bill bill = mock(PgeG11Bill.class);

        // WHEN
        sut.onListItemDismissed(position, bill);

        // THEN
        verify(view).setListData(any(LazyList.class));
        verify(view).onItemRemoveFromList(eq(position));
    }

    @Test
    public void onListItemDismissed_showsUndoDeleteMessage() throws Exception {
        // GIVEN
        final Bill bill = mock(PgeG11Bill.class);
        final int position = 0;

        // WHEN
        sut.onListItemDismissed(position, bill);

        // THEN
        verify(view).showUndoDeleteMessage(position);
    }

    @Test
    public void undoDeleteClicked_viewAddsItemToList() throws Exception {
        // WHEN
        final int position = 0;
        sut.undoDeleteClicked(position);

        // THEN
        verify(view).setListData(any(LazyList.class));
        verify(view).onItemInsertedToList(eq(position));
    }

    @Test
    public void undoDeleteClicked_undosDeleteHistory() throws Exception {
        // WHEN
        sut.undoDeleteClicked(0);

        // THEN
        verify(history).undoDelete();
    }

    @Test
    public void undoDeleteClicked_fetchAllHistory() throws Exception {
        // WHEN
        sut.undoDeleteClicked(0);

        // THEN
        verify(history).getAll();
    }

    @Test
    public void onListItemClicked_whenNotInSelectMode_opensBill() throws Exception {
        // GIVEN
        Bill bill = mock(PgeG11Bill.class);
        View itemView = mock(View.class);
        HistoryViewItem item = given_historyViewItem(bill, itemView);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(view).openBill(bill, itemView);
    }

    @Test
    public void onListItemClicked_whenNotInSelectMode_registerSavedBill() throws Exception {
        // GIVEN
        Bill bill = mock(PgeG11Bill.class);
        HistoryViewItem item = given_historyViewItem(bill);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(savedBillsRegistry).register(bill);
    }

    @Test
    public void onListItemClicked_andItemSelected_deselectsViewItem() throws Exception {
        // GIVEN
        int position = 2;
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.isSelected(position)).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(item).deselect();
    }

    @Test
    public void onListItemClicked_andItemSelected_deselectPosition() throws Exception {
        // GIVEN
        int position = 2;
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.isSelected(position)).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(selection).deselect(position);
    }

    @Test
    public void onListItemClicked_whenInSelectMode_andItemNotSelected_selectsViewItem() throws Exception {
        // GIVEN
        when(selection.isAnySelected()).thenReturn(true);
        int position = 2;
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(item).select();
    }

    @Test
    public void onListItemClicked_whenInSelectMode_andItemNotSelected_selectsPosition() throws Exception {
        // GIVEN
        when(selection.isAnySelected()).thenReturn(true);
        int position = 2;
        PgeG11Bill bill = mock(PgeG11Bill.class);
        HistoryViewItem item = given_historyViewItem(bill, mock(View.class), position);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(selection).select(position, bill);
    }

    @Test
    public void onListItemClicked_onSingleSelectedItem_hidesDeleteButton() throws Exception {
        // GIVEN
        when(selection.isAnySelected()).thenReturn(true, false);
        int position = 2;
        when(selection.isSelected(position)).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemClicked(item);

        // THEN
        verify(view).hideDeleteButton();
    }

    @Test
    public void onListItemLongClicked_whenNotInSelectMode_showsDeleteButton() throws Exception {
        // GIVEN
        HistoryViewItem item = given_historyViewItem(2);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(view).showDeleteButton();
    }

    @Test
    public void onListItemLongClicked_whenNotInSelectMode_selectsViewItem() throws Exception {
        // GIVEN
        HistoryViewItem item = given_historyViewItem(2);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(item).select();
    }

    @Test
    public void onListItemLongClicked_andItemSelected_deselectsViewItem() throws Exception {
        // GIVEN
        int position = 2;
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.isSelected(position)).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(item).deselect();
    }

    @Test
    public void onListItemLongClicked_andItemSelected_deselectsPosition() throws Exception {
        // GIVEN
        int position = 2;
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.isSelected(position)).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(selection).deselect(position);
    }

    @Test
    public void onListItemLongClicked_whenInSelectMode_andItemNotSelected_selectsViewItem() throws Exception {
        // GIVEN
        when(selection.isAnySelected()).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(2);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(item).select();
    }

    @Test
    public void onListItemLongClicked_whenItemNotSelected_selectsPosition() throws Exception {
        // GIVEN
        int position = 2;
        Bill bill = mock(Bill.class);
        HistoryViewItem item = given_historyViewItem(bill, mock(View.class), position);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(selection).select(position, bill);
    }

    @Test
    public void onListItemLongClicked_onSingleSelectedItem_hidesDeleteButton() throws Exception {
        // GIVEN
        int position = 2;
        when(selection.isAnySelected()).thenReturn(true, false);
        when(selection.isSelected(position)).thenReturn(true);
        HistoryViewItem item = given_historyViewItem(position);

        // WHEN
        sut.onListItemLongClicked(item);

        // THEN
        verify(view).hideDeleteButton();
    }

    @Test
    public void deleteClicked_deletesSelectedBill() throws Exception {
        // GIVEN
        final Long id = 1L;
        final Long pricesId = 2L;
        Bill bill = mock(PgeG11Bill.class);
        when(bill.getId()).thenReturn(id);
        when(bill.getPricesId()).thenReturn(pricesId);

        when(selection.isAnySelected()).thenReturn(true);
        when(selection.getItems()).thenReturn(Arrays.asList(bill, bill));
        when(selection.getPositionsReverseOrder()).thenReturn(new int[]{3, 2});

        // WHEN
        sut.deleteClicked();

        // THEN
        verify(history, times(2)).deleteBillWithPrices(BillType.PGE_G11, id, pricesId);
    }

    @Test
    public void deleteClicked_removesViewsFromList() throws Exception {
        // GIVEN
        int position1 = 2;
        int position2 = 5;
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.getPositionsReverseOrder()).thenReturn(new int[]{position2, position1});

        // WHEN
        sut.deleteClicked();

        // THEN
        verify(view).setListData(any(LazyList.class));
        verify(view).onItemRemoveFromList(eq(position2));
        verify(view).onItemRemoveFromList(eq(position1));
    }

    @Test
    public void deleteClicked_clearsSelection() throws Exception {
        // GIVEN
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.getPositionsReverseOrder()).thenReturn(new int[0]);

        // WHEN
        sut.deleteClicked();

        // THEN
        verify(selection).deselectAll();
    }

    @Test
    public void deleteClicked_hidesDeleteButton() throws Exception {
        // GIVEN
        when(selection.isAnySelected()).thenReturn(true);
        when(selection.getPositionsReverseOrder()).thenReturn(new int[0]);

        // WHEN
        sut.deleteClicked();

        // THEN
        verify(view).hideDeleteButton();
    }

    private HistoryViewItem given_historyViewItem(int position) {
        return given_historyViewItem(mock(PgeG11Bill.class), mock(View.class), position);
    }

    private HistoryViewItem given_historyViewItem(Bill bill) {
        return given_historyViewItem(bill, mock(View.class), 0);
    }

    private HistoryViewItem given_historyViewItem(Bill bill, View itemView) {
        return given_historyViewItem(bill, itemView, 0);
    }

    private HistoryViewItem given_historyViewItem(Bill bill, View itemView, int position) {
        HistoryViewItem item = mock(HistoryViewItem.class);
        when(item.getBill()).thenReturn(bill);
        when(item.getView()).thenReturn(itemView);
        when(item.getPositionOnList()).thenReturn(position);
        return item;
    }
}

File: app/src/test/java/pl/srw/billcalculator/history/HistoryPresenterTest.kt
package pl.srw.billcalculator.history

import android.view.View
import com.nhaarman.mockito_kotlin.inOrder
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import junitparams.JUnitParamsRunner
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import pl.srw.billcalculator.bill.SavedBillsRegistry
import pl.srw.billcalculator.data.ApplicationRepo
import pl.srw.billcalculator.data.bill.HistoryRepo
import pl.srw.billcalculator.db.Bill
import pl.srw.billcalculator.db.PgeG11Bill
import pl.srw.billcalculator.history.list.item.HistoryViewItem
import pl.srw.billcalculator.setState
import pl.srw.billcalculator.util.BillSelection
import java.util.Arrays

@RunWith(JUnitParamsRunner::class)
class HistoryPresenterTest {

    val view: HistoryPresenter.HistoryView = mock()
    val applicationRepo: ApplicationRepo = mock()
    val history: HistoryRepo = mock()
    val savedBillsRegistry: SavedBillsRegistry = mock()
    val selection: BillSelection = mock()

    val sut = HistoryPresenter(applicationRepo, history, savedBillsRegistry, selection)

    @Before
    fun setUp() {
        sut.setState("view", view)
    }

    @Test
    fun onFirstBind_whenFirstLaunch_showsWelcomeDialog() {
        // GIVEN
        whenever(applicationRepo.isFirstLaunch).thenReturn(true)

        // WHEN
        sut.onFirstBind()

        // THEN
        verify(view).showWelcomeDialog()
    }

    @Test
    fun onFirstBind_whenFirstLaunch_doesNotShowNewUIDialog() {
        // GIVEN
        whenever(applicationRepo.isFirstLaunch).thenReturn(true)

        // WHEN
        sut.onFirstBind()

        // THEN
        verify(view, never()).showNewUIDialog()
    }

    @Test
    fun onFirstBind_whenFirstLaunch_marksHelpShown() {
        // GIVEN
        whenever(applicationRepo.isFirstLaunch).thenReturn(true)

        // WHEN
        sut.onFirstBind()

        // THEN
        verify(applicationRepo).markHelpShown()
    }

    @Test
    fun onFirstBind_whenNotFirstLaunch_dontShowsWelcomeDialog() {
        // GIVEN
        whenever(applicationRepo.isFirstLaunch).thenReturn(false)

        // WHEN
        sut.onFirstBind()

        // THEN
        verify(view, never()).showWelcomeDialog()
    }

    @Test
    fun onFirstBind_whenNotFirstLaunch_andHelpWasNotShown_showsNewUIDialogAndMarkShown() {
        // GIVEN
        whenever(applicationRepo.isFirstLaunch).thenReturn(false)
        whenever(applicationRepo.wasHelpShown()).thenReturn(false)

        // WHEN
        sut.onFirstBind()

        // THEN
        verify(view).showNewUIDialog()
        verify(applicationRepo).markHelpShown()
    }

    @Test
    fun onFirstBind_whenNotFirstLaunch_andHelpWasMarkShown_doesNotShowNewUIDialog() {
        // GIVEN
        whenever(applicationRepo.isFirstLaunch).thenReturn(false)
        whenever(applicationRepo.wasHelpShown()).thenReturn(true)

        // WHEN
        sut.onFirstBind()

        // THEN
        verify(view, never()).showNewUIDialog()
    }

    @Test
    fun `disables swipe delete, on new view restore state, when any item selected`() {
        whenever(selection.isAnySelected()).thenReturn(true)

        sut.onNewViewRestoreState()

        verify(view).disableSwipeDelete()
    }

    @Test
    fun `unselects bills, when history changes`() {
        sut.onHistoryChanged()

        verify(selection).deselectAll()
    }

    @Test
    fun `hides delete button, when history changes`() {
        sut.onHistoryChanged()

        verify(view).hideDeleteButton()
    }

    @Test
    fun `enables swipe delete, when history changes`() {
        sut.onHistoryChanged()

        verify(view).enableSwipeDelete()
    }

    @Test
    fun helpMenuClicked_showsHelp() {
        // WHEN
        sut.helpMenuClicked()

        // THEN
        verify(view).showHelp()
    }

    @Test
    fun onListItemDismissed_deleteBillWithPrices() {
        // GIVEN
        val bill: PgeG11Bill = mock()

        // WHEN
        sut.onListItemDismissed(0, bill)

        // THEN
        verify(history).deleteBillWithPrices(bill)
    }

    @Test
    fun `on list item dismissed, cache item for undo possibility`() {
        // GIVEN
        val bill: PgeG11Bill = mock()

        // WHEN
        sut.onListItemDismissed(0, bill)

        // THEN
        verify(history).cacheBillForUndoDelete(bill)
    }

    @Test
    fun onListItemDismissed_showsUndoDeleteMessage() {
        // GIVEN
        val bill: PgeG11Bill = mock()
        val position = 0

        // WHEN
        sut.onListItemDismissed(position, bill)

        // THEN
        verify(view).showUndoDeleteMessage(position)
    }

    @Test
    fun `undo delete clicked, scroll to last inserted item`() {
        // GIVEN
        whenever(history.isUndoDeletePossible()).thenReturn(true)

        // WHEN
        sut.undoDeleteClicked(3, 2, 5)

        // THEN
        verify(view).scrollToPosition(5)
    }

    @Test
    fun undoDeleteClicked_undosDeleteHistory() {
        // GIVEN
        whenever(history.isUndoDeletePossible()).thenReturn(true)

        // WHEN
        sut.undoDeleteClicked(0)

        // THEN
        verify(history).undoDelete()
    }

    @Test
    fun undoDeleteClicked_whenAlreadyUndo_doesNotUndoAnyMore() {
        // GIVEN
        whenever(history.isUndoDeletePossible()).thenReturn(false)

        // WHEN
        sut.undoDeleteClicked(0)

        // THEN
        verify(history).isUndoDeletePossible()
        verify(history, never()).undoDelete()
    }

    @Test
    fun `updates selection, after undo on multiple items`() {
        whenever(history.isUndoDeletePossible()).thenReturn(true)

        sut.undoDeleteClicked(1, 3, 2)

        inOrder(selection) {
            verify(selection).onInsert(1)
            verify(selection).onInsert(2)
            verify(selection).onInsert(3)
        }
    }

    @Test
    fun `updates selection, after undo clicked, before actual data changed`() {
        whenever(history.isUndoDeletePossible()).thenReturn(true)
        val lowestPosition = 1

        sut.undoDeleteClicked(lowestPosition)

        inOrder(selection, view, history) {
            verify(selection).onInsert(lowestPosition)
            verify(history).undoDelete()
            verify(view).scrollToPosition(lowestPosition)
        }
    }

    @Test
    fun onListItemClicked_whenNotInSelectMode_opensBill() {
        // GIVEN
        val bill: PgeG11Bill = mock()
        val itemView: View = mock()
        val item = given_historyViewItem(bill, itemView)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(view).openBill(bill, itemView)
    }

    @Test
    fun onListItemClicked_whenNotInSelectMode_registerSavedBill() {
        // GIVEN
        val bill: PgeG11Bill = mock()
        val item = given_historyViewItem(bill)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(savedBillsRegistry).register(bill)
    }

    @Test
    fun onListItemClicked_andItemSelected_deselectsViewItem() {
        // GIVEN
        val position = 2
        whenever(selection.isAnySelected()).thenReturn(true)
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(item).deselect()
    }

    @Test
    fun onListItemClicked_andItemSelected_deselectPosition() {
        // GIVEN
        val position = 2
        whenever(selection.isAnySelected()).thenReturn(true)
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(selection).deselect(position)
    }

    @Test
    fun onListItemClicked_whenInSelectMode_andItemNotSelected_selectsViewItem() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true)
        val position = 2
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(item).select()
    }

    @Test
    fun onListItemClicked_whenInSelectMode_andItemNotSelected_selectsPosition() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true)
        val position = 2
        val bill: PgeG11Bill = mock()
        val item = given_historyViewItem(bill, mock<View>(), position)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(selection).select(position, bill)
    }

    @Test
    fun onListItemClicked_onSingleSelectedItem_hidesDeleteButton() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true, false)
        val position = 2
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(view).hideDeleteButton()
    }

    @Test
    fun onListItemClicked_onSingleSelectedItem_enablesSwipeDelete() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true, false)
        val position = 2
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemClicked(item)

        // THEN
        verify(view).enableSwipeDelete()
    }

    @Test
    fun onListItemLongClicked_whenNotInSelectMode_showsDeleteButton() {
        // GIVEN
        val item = given_historyViewItem(position = 2)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(view).showDeleteButton()
    }

    @Test
    fun onListItemLongClicked_whenNotInSelectMode_disablesSwipeDelete() {
        // GIVEN
        val item = given_historyViewItem(position = 2)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(view).disableSwipeDelete()
    }

    @Test
    fun onListItemLongClicked_whenNotInSelectMode_selectsViewItem() {
        // GIVEN
        val item = given_historyViewItem(position = 2)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(item).select()
    }

    @Test
    fun onListItemLongClicked_andItemSelected_deselectsViewItem() {
        // GIVEN
        val position = 2
        whenever(selection.isAnySelected()).thenReturn(true)
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(item).deselect()
    }

    @Test
    fun onListItemLongClicked_andItemSelected_deselectsPosition() {
        // GIVEN
        val position = 2
        whenever(selection.isAnySelected()).thenReturn(true)
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(selection).deselect(position)
    }

    @Test
    fun onListItemLongClicked_whenInSelectMode_andItemNotSelected_selectsViewItem() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true)
        val item = given_historyViewItem(position = 2)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(item).select()
    }

    @Test
    fun onListItemLongClicked_whenItemNotSelected_selectsPosition() {
        // GIVEN
        val position = 2
        val bill: Bill = mock()
        val item = given_historyViewItem(bill, mock<View>(), position)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(selection).select(position, bill)
    }

    @Test
    fun onListItemLongClicked_onSingleSelectedItem_hidesDeleteButton() {
        // GIVEN
        val position = 2
        whenever(selection.isAnySelected()).thenReturn(true, false)
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(view).hideDeleteButton()
    }

    @Test
    fun onListItemLongClicked_onSingleSelectedItem_enablesSwipeDelete() {
        // GIVEN
        val position = 2
        whenever(selection.isAnySelected()).thenReturn(true, false)
        whenever(selection.isSelected(position)).thenReturn(true)
        val item = given_historyViewItem(position = position)

        // WHEN
        sut.onListItemLongClicked(item)

        // THEN
        verify(view).enableSwipeDelete()
    }

    @Test
    fun `delete button clicked, caches bills for undo possibility`() {
        // GIVEN
        val items = listOf(mock<Bill>(), mock<Bill>())
        whenever(selection.getItems()).thenReturn(items)
        whenever(selection.getPositionsReverseOrder()).thenReturn(intArrayOf(1, 0))

        // WHEN
        sut.deleteClicked()

        // THEN
        verify(history).cacheBillsForUndoDelete(items)
    }

    @Test
    fun deleteClicked_deletesSelectedBills() {
        // GIVEN
        val bill: PgeG11Bill = mock()
        val bills = Arrays.asList<Bill>(bill, bill)
        whenever(selection.getItems()).thenReturn(bills)
        whenever(selection.getPositionsReverseOrder()).thenReturn(intArrayOf(3, 2))

        // WHEN
        sut.deleteClicked()

        // THEN
        verify(history).deleteBillsWithPrices(bills)
    }

    @Test
    fun `delete button clicked, enables swipe on items`() {
        // GIVEN
        whenever(selection.getPositionsReverseOrder()).thenReturn(intArrayOf(1))

        // WHEN
        sut.deleteClicked()

        // THEN
        verify(view).enableSwipeDelete()
    }

    @Test
    fun `delete button clicked, shows undo delete message`() {
        // GIVEN
        whenever(selection.getPositionsReverseOrder()).thenReturn(intArrayOf(5, 3))

        // WHEN
        sut.deleteClicked()

        // THEN
        verify(view).showUndoDeleteMessage(5, 3)
    }

    @Test
    fun deleteClicked_clearsSelection() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true)
        whenever(selection.getPositionsReverseOrder()).thenReturn(IntArray(0))

        // WHEN
        sut.deleteClicked()

        // THEN
        verify(selection).deselectAll()
    }

    @Test
    fun deleteClicked_hidesDeleteButton() {
        // GIVEN
        whenever(selection.isAnySelected()).thenReturn(true)
        whenever(selection.getPositionsReverseOrder()).thenReturn(IntArray(0))

        // WHEN
        sut.deleteClicked()

        // THEN
        verify(view).hideDeleteButton()
    }

    private fun given_historyViewItem(bill: Bill = mock<PgeG11Bill>(), itemView: View = mock<View>(), position: Int = 0): HistoryViewItem {
        val item = mock<HistoryViewItem>()
        whenever(item.bill).thenReturn(bill)
        whenever(item.view).thenReturn(itemView)
        whenever(item.positionOnList).thenReturn(position)
        return item
    }
}
